🌱 Fix self-hosted flakes in E2E tests

fabriziopandini opened the pull request:
**What this PR does / why we need it**:
This PR aims to fix self-hosted flakes by ensuring that mhc is not run during this test and by adding two delay in order to avoid doing move too aggressively.

In order to make this happen the entire management of cluster templates was re-architected so:
- cluster templates for tests are now generated from a set composable bases
- MHC-remediation has a new dedicated template
- all the other templates are without MHC
- also, in order to simplify things, this PR drops the distinction between CI and DEV test config

On top of that, this PR includes, a set of nits for improving test logs (mostly uppercase)

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3589

fabriziopandini commented:
/milestone v0.3.10

fabriziopandini commented:
/retest

fabriziopandini commented on test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:
```diff
@@ -0,0 +1,70 @@
+---
```
I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

fabriziopandini commented on test/e2e/data/infrastructure-docker/cluster-template-kcp-adoption.yaml:
```diff
@@ -1,179 +1,180 @@
-###
-apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
-kind: DockerCluster
+apiVersion: v1
```
I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

fabriziopandini commented on test/e2e/data/infrastructure-docker/cluster-template-mhc.yaml:
```diff
@@ -1,147 +1,154 @@
----
-kind: KubeadmControlPlane
-apiVersion: controlplane.cluster.x-k8s.io/v1alpha3
+apiVersion: v1
```
I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

fabriziopandini commented on test/e2e/data/infrastructure-docker/cluster-template.yaml:
```diff
@@ -1,147 +1,139 @@
-apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
-kind: DockerCluster
+apiVersion: v1
```
I would like to gitignore the generated cluster-templates and keep under source control only the bases, but it seems that this makes the apidiff batch to fail 🤔

fabriziopandini commented on test/e2e/config/docker-dev.yaml:
```diff
@@ -1,123 +0,0 @@
----
-# E2E test scenario using local dev images and manifests built from the source tree for following providers:
```
I'm dropping dev config, so we should not anymore put efforts in keeping it up-to-date

fabriziopandini commented on test/e2e/data/infrastructure-docker/metadata.yaml:
```diff
@@ -1,9 +0,0 @@
-apiVersion: clusterctl.cluster.x-k8s.io/v1alpha3
```
Dropping this file because now it is embedded in clusterctl

fabriziopandini commented on test/e2e/kcp_adoption.go:
```diff
@@ -131,21 +132,24 @@ func KCPAdoptionSpec(ctx context.Context, inputGetter func() KCPAdoptionSpecInpu
+		Expect(input.BootstrapClusterProxy.ApplyWithArgs(ctx, workloadClusterTemplate, "--selector", "kcp-adoption.step2")).ShouldNot(HaveOccurred())
+
+		var controlPlane *controlplanev1.KubeadmControlPlane
+		Eventually(func() *controlplanev1.KubeadmControlPlane {
```
Adding Eventually to fix some flakes observed during local tests

fabriziopandini commented:
/hold 
investigating an error on ci-e2e.sh that shows up un CI only, while locally everything works just fine 🤔

vincepri commented on scripts/ci-e2e.sh:
```diff
@@ -31,7 +31,7 @@ source "${REPO_ROOT}/hack/ensure-kustomize.sh"
 # Configure provider images generation;
 # please ensure the generated image name matches image names used in the E2E_CONF_FILE
 export REGISTRY=gcr.io/k8s-staging-cluster-api
-export TAG=ci
```
Why change the tag for the CI?

fabriziopandini replied:
moved to https://github.com/kubernetes-sigs/cluster-api/pull/3651

vincepri commented on test/e2e/Makefile:
```diff
@@ -28,7 +28,7 @@ export GOPROXY
 REPO_ROOT := $(shell git rev-parse --show-toplevel)
 
 help:  ## Display this help
-	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
```
Was this change intentional? What does it do?

fabriziopandini replied:
moved to #3651

vincepri commented:
From the PR's description, seems we're now splitting manifests in order to separate MHC components? Would this be something that we would expect users to do as well? 

During move, we should be able to pause controllers and look at the state of the cluster if it's safe to move, is there anything stopping us to do so?

Can we split this PR in multiple ones? In particular:
> also, in order to simplify things, this PR drops the distinction between CI and DEV test config
> On top of that, this PR includes, a set of nits for improving test logs (mostly uppercase)

These changes seems outside of the PR's scope

sedefsavas commented on test/e2e/config/docker.yaml:
```diff
@@ -5,17 +5,17 @@
 # - docker
 
-# For creating local dev images run ./scripts/ci-e2e.sh
+# For creating local dev images run make docker-build-e2e from the main CAPI repository
```
I think we can keep `run ./scripts/ci-e2e.sh` here, because running `make docker-build-e2e` will not create images with the tag `dev` this file is using.

fabriziopandini replied:
moved to #3651

sedefsavas commented on test/e2e/data/infrastructure-docker/cluster-template/kustomization.yaml:
```diff
@@ -0,0 +1,5 @@
+bases:
```
nit. We can generate cluster-template.yaml on the fly and use that as a base to avoid extra changes we will need to do here when apis change.

fabriziopandini replied:
moved to #3651

sedefsavas commented on test/e2e/Makefile:
```diff
@@ -38,28 +38,43 @@ TOOLS_DIR := $(REPO_ROOT)/hack/tools
 ## --------------------------------------
 
-TEST_E2E_DIR := $(REPO_ROOT)/test/e2e
+DOCKER_TEMPLATES := $(REPO_ROOT)/test/e2e/data/infrastructure-docker
```
Can we put them under `templates` dir like other providers? kcp-adoption does not make much sense to put it under that though.

If we will keep them under `e2e/data` then we can modify `cluster-templates` below to `e2e-cluster-templates` to be explicit about their purpose.

fabriziopandini replied:
moved to #3651

sedefsavas commented on test/framework/machine_helpers.go:
```diff
@@ -120,8 +121,7 @@ func WaitForControlPlaneMachinesToBeUpgraded(ctx context.Context, input WaitForC
 
-	By("ensuring all machines have upgraded kubernetes version")
-	log.Logf("Ensuring all MachineDeployment Machines have upgraded kubernetes version %s", input.KubernetesUpgradeVersion)
+	By(fmt.Sprintf("Ensuring all control-plane have upgraded kubernetes version %s", input.KubernetesUpgradeVersion))
```
```suggestion
	By(fmt.Sprintf("Ensuring all control-plane machines have upgraded kubernetes version %s", input.KubernetesUpgradeVersion))
```

fabriziopandini replied:
fixed in https://github.com/kubernetes-sigs/cluster-api/pull/3649

sedefsavas reviewed (commented):
Thanks for the fix!

This PR is also addressing this issue: https://github.com/kubernetes-sigs/cluster-api/issues/3461
I am fine with adding it in the title and the issues fixed by this PR or making this separate PRs.

sedefsavas commented:
@fabriziopandini Locally, works for me as well.


/test pull-cluster-api-e2e

sedefsavas commented:
In `test/e2e/Makefile`

    cd $(TOOLS_DIR); go build -tags=tools -o $(BIN_DIR)/kustomize sigs.k8s.io/kustomize/kustomize/v3

should be 

	cd $(TOOLS_DIR) && go build -tags=tools -o $(BIN_DIR)/kustomize sigs.k8s.io/kustomize/kustomize/v3

fabriziopandini commented:
@vincepri 

> From the PR's description, seems we're now splitting manifests in order to separate MHC components? Would this be something that we would expect users to do as well?

No. The current MHC object in E2E tests is specifically configured to always trigger remediation after 30s the node is started, and I removed it from the other e2e tests in order to avoid flakes due to remediation kicking in case of slow execution

> During move, we should be able to pause controllers and look at the state of the cluster if it's safe to move, is there anything stopping us to do so?

This is not how the move logics works. Currently we are pausing reconciliation on clusters included in the Move scope, but the controller will continue to run.

> Can we split this PR in multiple ones? 

Done! rif 
https://github.com/kubernetes-sigs/cluster-api/pull/3649
https://github.com/kubernetes-sigs/cluster-api/pull/3650
https://github.com/kubernetes-sigs/cluster-api/pull/3651

> These changes seems outside of the PR's scope

Some of them are nits, but refactoring templates for MHC is potentially related to self-hosted flakes, see the note above about MHC and https://github.com/kubernetes-sigs/cluster-api/issues/3589#issuecomment-688333335.
However, no problem to move to another PR

fabriziopandini commented:
/hold cancel
/test pull-cluster-api-e2e-full

vincepri commented:
> No. The current MHC object in E2E tests is specifically configured to always trigger remediation after 30s the node is started, and I removed it from the other e2e tests in order to avoid flakes due to remediation kicking in case of slow execution

I assume this change has moved to the other PR, I'll take a look there before commenting

> This is not how the move logics works. Currently we are pausing reconciliation on clusters included in the Move scope, but the controller will continue to run.

We should probably consider stopping all controllers before move, for safety purposes and to avoid any running slow worker to perform actions while move is running.

vincepri commented:
/test pull-cluster-api-e2e-full

randomvariable commented:
/lgtm

pending e2e pass

vincepri reviewed (approved):
/approve
/lgtm
